### PR TITLE
p2p message parsing

### PR DESCRIPTION
**Pull request by anonymoussprocket**

### CODE REVIEW

File: src/tezos/TezosOperations.ts [hunk 1]
```diff
@@ -14,16 +14,16 @@ import {Operation} from "./TezosTypes";
...
 }
 
 /**
  * Result of a successfully sent operation
  */
 export interface OperationResult {
-    results: TezosTypes.AlphaOperationsWithMetadata,
-    operationGroupID: String
+    results: TezosTypes.AlphaOperationsWithMetadata;
+    operationGroupID: string;
```
Comment (vishakh) [inline src/tezos/TezosOperations.ts]: In the past I have seen variable behaviour depending on whether one uses `string` or `String`. Are we sure no behaviour has changed here?

File: src/tezos/TezosOperations.ts [hunk 2]
```diff
@@ -83,11 +83,22 @@ export namespace TezosOperations {
...
-                branch: blockHead.hash,
-                contents: operations
-            };
-        return TezosNode.forgeOperation(network, payload)
+        const payload = { branch: blockHead.hash, contents: operations };
+        const ops = await TezosNode.forgeOperation(network, payload);
+
+        const decoded = TezosMessageCodec.parseOperationGroup(ops);
+
+        for (let i = 0; i < operations.length; i++) {
```
Comment (vishakh) [inline src/tezos/TezosOperations.ts]: Is there a more canonical iteration we can use? I think a `for` loop like this is very rarely used in JS / TS.

File: src/tezos/TezosOperations.ts [hunk 3]
```diff
@@ -83,11 +83,22 @@ export namespace TezosOperations {
...
+            const clientop = operations[i];
+            if (clientop["kind"] !== "transaction") { continue; }
```
Comment (vishakh) [inline src/tezos/TezosOperations.ts]: How come we only validate transactions?

File: src/tezos/TezosMessageCodec.ts [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+    return envelope;
+  }
+
+  /**
+   * Parse a reveal message possibly containing siblings.
+   * @param {string} revealMessage Encoded reveal-type message
+   * @param {boolean} isFirst Flag to indicate first operation of Operation Group.
+   */
+  export function parseReveal(revealMessage, isFirst = true): OperationEnvelope {
```
Comment (vishakh) [inline src/tezos/TezosMessageCodec.ts]: What are the types of these parameters?

File: src/tezos/TezosMessageCodec.ts [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+  "accountActivate",
+  "proposal",
+  "ballot",
+  "reveal",
+  "transaction",
+  "origination",
+  "delegation"
+];
+
+export namespace TezosMessageCodec {
```
Comment (vishakh) [inline src/tezos/TezosMessageCodec.ts]: Could we leave a comment about which kinds of functions this namespace contains?

File: src/tezos/TezosMessageUtil.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
+/*eslint no-bitwise: 0*/
+import base58check from "bs58check";
+import baseN from "base-n";
+
+const base128 = baseN.create({
+  characters: [...Array(128).keys()].map(k => ("0" + k.toString(16)).slice(-2))
+});
+
+export namespace TezosMessageUtils {
```
Comment (vishakh) [inline src/tezos/TezosMessageUtil.ts]: Could we leave a comment about which kinds of functions this namespace contains?

File: integration_test/TezosMessageCodec.spec.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import { TezosOperations } from "../src";
+import { TezosMessageCodec } from "../src/tezos/TezosMessageCodec";
+import "mocha";
+import { servers } from "../test/servers";
+import { TezosWallet } from "../src";
+import { KeyStore } from "../src/types/KeyStore";
+
+const tezosURL = servers.tezosServer;
+
+//Software tezos operations do not require a valid derivation path
```
Comment (vishakh) [inline integration_test/TezosMessageCodec.spec.ts]: Could we get rid of the dead code?

File: integration_test/TezosMessageCodec.spec.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+//     'efcvoykz.kygxsosz@tezos.example.org',
+//     'BFdu9bDJxC'
+// );
+
+describe("signOperationGroup() and computeOperationHash()", () => {
+  it("correctly forges a transaction from binary", () => {
+    let forgedTransaction =
+      "9cd3aa823667cd8223073d2447c4c012373c99a00df5112b502ee82f5cf95ae408000069ef8fb5d47d8a4321c94576a2316a632be8ce898094ebdc039cfe19bc50ac0280a0c21e00004783ca75ac8736021036aa5ff2a69f116eca759800";
+    const result = TezosMessageCodec.parseTransaction(forgedTransaction);
+    expect(result.type).to.equal("transaction");
```
Comment (vishakh) [inline integration_test/TezosMessageCodec.spec.ts]: The `OperationEnvelope` interface does not seem to contain many of these `result` members.

```typescript
interface OperationEnvelope {
    operation: Operation,
    branch: string,
    next?: string,
    nextoffset: number
  }
```

File: integration_test/TezosMessageCodec.spec.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+const invalidDerivationPath = "ighiehgieh";
+
+// const keys = <KeyStore> TezosWallet.unlockFundraiserIdentity(
+//     'bomb sing vacant repair illegal category unveil color olive chest wink expand fringe pioneer reward',
+//     'efcvoykz.kygxsosz@tezos.example.org',
+//     'BFdu9bDJxC'
+// );
+
+describe("signOperationGroup() and computeOperationHash()", () => {
+  it("correctly forges a transaction from binary", () => {
```
Comment (vishakh) [inline integration_test/TezosMessageCodec.spec.ts]: In these tests, would the word 'parses' make more sense than 'forges'?

File: integration_test/TezosMessageCodec.spec.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+//Software tezos operations do not require a valid derivation path
+const invalidDerivationPath = "ighiehgieh";
+
+// const keys = <KeyStore> TezosWallet.unlockFundraiserIdentity(
+//     'bomb sing vacant repair illegal category unveil color olive chest wink expand fringe pioneer reward',
+//     'efcvoykz.kygxsosz@tezos.example.org',
+//     'BFdu9bDJxC'
+// );
+
+describe("signOperationGroup() and computeOperationHash()", () => {
```
Comment (vishakh) [inline integration_test/TezosMessageCodec.spec.ts]: Could we test one operation of each kind we currently can parse? Right now, we are testing only two.

**Comment by vishakh** [comment]
Is it intentional that we are checking in changes to the `dist` folder?

**Comment by Sm00g15** [comment]
> Is it intentional that we are checking in changes to the `dist` folder?

Unintentional, changing that now and pushing up

File: src/tezos/TezosOperations.ts (hunk 1 above)
Comment (Sm00g15) [reply src/tezos/TezosOperations.ts]: https://stackoverflow.com/questions/14727044/typescript-difference-between-string-and-string

File: src/tezos/TezosMessageCodec.ts (hunk 4 above)
Comment (Sm00g15) [reply src/tezos/TezosMessageCodec.ts]: `export function parseReveal(revealMessage: string, isFirst:boolean = true)`

File: src/tezos/TezosOperations.ts (hunk 2 above)
Comment (Sm00g15) [reply src/tezos/TezosOperations.ts]:
```
    operations.forEach(clientop => {
      decoded.forEach(serverop => {
        if (clientop["kind"] !== "transaction") {
          if (
            serverop.kind !== clientop["kind"] ||
            serverop.fee !== clientop["fee"] ||
            serverop.amount !== clientop["amount"] ||
            serverop.destination !== clientop["destination"]
          ) {
            return ops;
          }
        }
      });
    });
``` I think that'd do the trick

**Pull request merged**